src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -144,11 +174,16 @@ public async Task<bool> OnDocumentClosedAsync(TextDocument document, IGlobalOpti
             {
                 // can not be cancelled
                 // remove active file state and put it in project state
-                if (!_activeFileStates.TryRemove(document.Id, out var activeFileState))
+                if (!_activeFileStates.TryGetValue(document.Id, out var activeFileState))
                 {
                     return false;
                 }
 
+                lock (_lock)
+                {
+                    _activeFileStates = _activeFileStates.Remove(document.Id);
+                }
```

Comment:
this feels racey with the above as well.  the check seems like it shuld be in the lock as well.

note: the prior code also seemed to suffer from this... so not sure how much i absolutely care.

Reply:
As you mentioned, the On* methods are called much less frequently, so just moved the initial checks inside the lock.

Reply:
Yep, On* methods are all called only for Solution crawler push diagnostics mode (as the On prefix also suggests). These won't be called concurrently, but adding the lock in these callbacks is also fine as they are not invoked very frequently. Some day we can clean this all up.